src/python/pants/backend/native/subsystems/binaries/gcc.py

```diff
@@ -7,64 +7,81 @@
-  def subsystem_dependencies(cls):
-    return super(GCC, cls).subsystem_dependencies() + (ParseSearchDirs.scoped(cls),)
-
-  @memoized_property
-  def _parse_search_dirs(self):
-    return ParseSearchDirs.scoped_instance(self)
-
-  def _lib_search_dirs(self, compiler_exe, path_entries):
-    return self._parse_search_dirs.get_compiler_library_dirs(
-      compiler_exe,
-      env={'PATH': create_path_env_var(path_entries)})
-
   @memoized_method
   def path_entries(self):
     return [os.path.join(self.select(), 'bin')]
 
-  def c_compiler(self):
-    exe_filename = 'gcc'
-    path_entries = self.path_entries()
+  _PLATFORM_INTERMEDIATE_DIRNAME = {
+    'darwin': lambda: 'x86_64-apple-darwin17.5.0',
+    'linux': lambda: 'x86_64-pc-linux-gnu',
+  }
+
+  @memoized_method
+  def _common_lib_dirs(self, platform):
+    return [
+      os.path.join(self.select(), 'lib'),
+      os.path.join(self.select(), 'lib64'),
+      os.path.join(self.select(), 'lib/gcc'),
+      os.path.join(self.select(), 'lib/gcc',
+                   platform.resolve_platform_specific(self._PLATFORM_INTERMEDIATE_DIRNAME),
+                   self.version()),
+    ]
+
+  @memoized_method
+  def _common_include_dirs(self, platform):
+    return [
+      os.path.join(self.select(), 'include'),
+      os.path.join(self.select(), 'lib/gcc',
+                   platform.resolve_platform_specific(self._PLATFORM_INTERMEDIATE_DIRNAME),
+                   self.version(),
+                   'include'),
+    ]
+
+  def c_compiler(self, platform):
     return CCompiler(
-      path_entries=path_entries,
-      exe_filename=exe_filename,
-      library_dirs=self._lib_search_dirs(exe_filename, path_entries),
-      include_dirs=[])
-
-  def cpp_compiler(self):
-    exe_filename = 'g++'
-    path_entries = self.path_entries()
+      path_entries=self.path_entries(),
+      exe_filename='gcc',
+      library_dirs=self._common_lib_dirs(platform),
+      include_dirs=self._common_include_dirs(platform),
+      extra_args=[])
+
+  @memoized_method
+  def _cpp_include_dirs(self, platform):
+    return [
+      os.path.join(self.select(), 'include/c++', self.version()),
+      os.path.join(self.select(), 'include/c++', self.version(),
+                   platform.resolve_platform_specific(self._PLATFORM_INTERMEDIATE_DIRNAME)),
+    ]
+
+  def cpp_compiler(self, platform):
     return CppCompiler(
       path_entries=self.path_entries(),
-      exe_filename=exe_filename,
-      library_dirs=self._lib_search_dirs(exe_filename, path_entries),
-      include_dirs=[])
+      exe_filename='g++',
+      library_dirs=self._common_lib_dirs(platform),
+      include_dirs=(self._common_include_dirs(platform) + self._cpp_include_dirs(platform)),
+      extra_args=[])
 
 
-@rule(GCCCCompiler, [Select(GCC)])
-def get_gcc(gcc):
-  yield GCCCCompiler(gcc.c_compiler())
+@rule(GCCCCompiler, [Select(GCC), Select(Platform)])
+def get_gcc(gcc, platform):
+  yield GCCCCompiler(gcc.c_compiler(platform))
```

Comment:
Why a yield here and below instead of a return? Afaict that just causes a little extra work looping through generator send gymnastics for what will be a blocking call either way: https://github.com/pantsbuild/pants/blob/b42d30524c2fb2823f2c05505a02326fbfe9a0be/src/rust/engine/src/nodes.rs#L861-L871

Reply:
It had slipped my mind that `return` was allowed for an `@rule` -- this is using `return` now.

Reply:
It absolutely makes sense, but I wasn't aware before that `yield` was interpreted to mean this (hadn't thought about it in a while) -- this makes the execution process of `@rule`s more clear to me.